src/client.rs

```diff
@@ -285,71 +280,181 @@ impl Client {
-        // Note: when we implement the `sync()` method in #28,
-        // we will remove this, and `.get()` will return the current content of cache if it
-        // is not empty.
-        let remote_timestamp = get_latest_change_timestamp(
-            &self.server_url,
-            &self.bucket_name,
-            &self.collection_name,
-        )?;
+        let remote_timestamp = match expected.into() {
+            Some(v) => v,
+            None => {
+                debug!("Obtain current timestamp.");
+                get_latest_change_timestamp(
+                    &self.server_url,
+                    &self.bucket_name,
+                    &self.collection_name,
+                )?
+            }
+        };
 
-        if let Some(collection) = stored {
+        if let Some(ref collection) = stored {
             let up_to_date = collection.timestamp == remote_timestamp;
             if up_to_date && self.verifier.verify(&collection).is_ok() {
                 debug!("Local data is up-to-date and valid.");
-                return Ok(collection.records);
+                return Ok(collection.to_owned());
             }
         }
 
         info!("Local data is empty, outdated, or has been tampered. Fetch from server.");
-        let collection = self._fetch_records_and_verify(remote_timestamp)?;
+        let (local_records, local_timestamp) = match stored {
+            Some(c) => (c.records, Some(c.timestamp)),
+            None => (Vec::new(), None),
+        };
+
+        let changeset = get_changeset(
+            &self.server_url,
+            &self.bucket_name,
+            &self.collection_name,
+            Some(remote_timestamp),
+            local_timestamp,
+        )?;
+
+        debug!(
+            "Apply {} changes to {} local records",
+            changeset.changes.len(),
+            local_records.len()
+        );
+        let merged = merge_changes(local_records.to_vec(), changeset.changes);
+
+        let collection = Collection {
+            bid: self.bucket_name.to_owned(),
+            cid: self.collection_name.to_owned(),
+            metadata: changeset.metadata,
+            records: merged,
+            timestamp: changeset.timestamp,
+        };
+
+        debug!("Verify signature after merge of changes with previous local data.");
+        self.verifier.verify(&collection)?;
+
+        debug!("Store collection with key={:?}", storage_key);
         let collection_bytes: Vec<u8> = serde_json::to_string(&collection)?.into();
         self.storage.store(&storage_key, collection_bytes)?;
-        Ok(collection.records)
+
+        Ok(collection)
+    }
+}
+
+fn merge_changes(
+    local_records: Vec<KintoObject>,
+    remote_changes: Vec<KintoObject>,
+) -> Vec<KintoObject> {
+    // Merge changes by record id and delete tombstones.
+    let mut local_by_id: HashMap<String, KintoObject> = HashMap::new();
+    for record in local_records {
+        local_by_id.insert(record["id"].to_string(), record);
+    }
```

Comment:
Another common way of writing this:
```suggestion
    let mut local_by_id: HashMap<String, KintoObject> = local_records
        .into_iter()
        .map(|record| (record["id"].to_string(), record))
        .collect();
```
This version sometimes allows the compiler to optimize the code better, e.g. by only requiring a single allocation, since the size of the `HashMap` is known from the get-go. Dynamically growing a `HashMap` generally takes multiple allocations.

The id is serialized to JSON again here. I don't think that's a problem – it just feels a bit unnatural to me.